server/src/main/java/io/druid/curator/discovery/CuratorDruidNodeDiscoveryProvider.java

```diff
@@ -0,0 +1,327 @@
+          }
+        });
+      }
+
+      nodeListeners.add(listener);
+    }
+
+    @Override
+    public synchronized void childEvent(CuratorFramework client, PathChildrenCacheEvent event) throws Exception
+    {
+      try {
+        switch (event.getType()) {
+          case CHILD_ADDED: {
+            final byte[] data;
+            try {
+              data = curatorFramework.getData().decompressed().forPath(event.getData().getPath());
+            }
+            catch (Exception ex) {
+              log.error(ex, "Failed to get data for path [%s]. Ignoring event [%s].", event.getData().getPath(), event.getType());
+              return;
+            }
+
+            DiscoveryDruidNode druidNode = jsonMapper.readValue(
+                data,
+                DiscoveryDruidNode.class
+            );
+
+            if (!nodeType.equals(druidNode.getNodeType())) {
+              log.error(
+                  "Node[%s:%s] add is discovered by node watcher of nodeType [%s]. Ignored.",
+                  druidNode.getNodeType(),
+                  druidNode,
+                  nodeType
+              );
+              return;
+            }
+
+            log.info("Received event [%s] for Node[%s:%s].", event.getType(), druidNode.getNodeType(), druidNode);
+
+            addNode(druidNode);
+
+            break;
+          }
+          case CHILD_REMOVED: {
+            DiscoveryDruidNode druidNode = jsonMapper.readValue(event.getData().getData(), DiscoveryDruidNode.class);
+
+            if (!nodeType.equals(druidNode.getNodeType())) {
+              log.error(
+                  "Node[%s:%s] removal is discovered by node watcher of nodeType [%s]. Ignored.",
+                  druidNode.getNodeType(),
+                  druidNode,
+                  nodeType
+              );
+              return;
+            }
+
+            log.info("Node[%s:%s] disappeared.", druidNode.getNodeType(), druidNode);
+
+            removeNode(druidNode);
+
+            break;
+          }
+          default: {
+            log.error("Ignored event type [%s] for nodeType [%s] watcher.", event.getType(), nodeType);
+          }
+        }
+      }
+      catch (Exception ex) {
+        log.error(ex, "unknown error in node watcher for type [%s].", nodeType);
+      }
+    }
+
+    private synchronized void addNode(DiscoveryDruidNode druidNode)
+    {
+      DiscoveryDruidNode prev = nodes.put(druidNode.getDruidNode().getHostAndPortToUse(), druidNode);
+      if (prev == null) {
+        for (DruidNodeDiscovery.Listener l : nodeListeners) {
+          listenerExecutor.submit(() -> {
+            try {
+              l.nodeAdded(druidNode);
+            }
+            catch (Exception ex) {
+              log.error(ex, "Exception occured in DiscoveryDruidNode.nodeAdded(node=[%s]) in listener [%s].", druidNode, l);
+            }
+          });
+        }
+      } else {
+        log.warn("Node[%s] discovered but existed already [%s].", druidNode, prev);
```

Comment:
In this case, the old state will be wiped out of the map.  I haven't done the homework to double check, but is it ok that the old state is wiped out, or might there have been state updates that would make us want to keep the old object and throw away the new one instead?